### PR TITLE
Fix for #98 - cache cannot be used with ADFS

### LINKED ISSUE #98: MsalCacheHelper.CreateAsync throws MsalClientException when using an ADFS ClientId
When trying to use `MsalCacheHelper` with a ClientId for ADFS which is not a GUID the followoing exception is thrown.

```
MSAL.NetCore.4.16.1.0.MsalClientException:
ErrorCode: client_id_must_be_guid
Microsoft.Identity.Client.MsalClientException: Error: ClientId is not a Guid.
  at Microsoft.Identity.Client.AbstractApplicationBuilder1.Validate() 
  at Microsoft.Identity.Client.PublicClientApplicationBuilder.Validate() 
  at Microsoft.Identity.Client.Extensions.Msal.MsalCacheHelper.GetAccountIdentifiersAsync(StorageCreationProperties storageCreationProperties, TraceSourceLogger logger) 
  at Microsoft.Identity.Client.Extensions.Msal.MsalCacheHelper.CreateAsync(StorageCreationProperties storageCreationProperties, TraceSource logger)
```

To reproduce this you can run the following.

``` c#
StorageCreationProperties storageProperties = new StorageCreationPropertiesBuilder("msal.cache", ".", "AdfsNonGuidClientId").Build();

await MsalCacheHelper.CreateAsync(storageProperties).ConfigureAwait(false);
```

To work around this issue I'm updating the code to supply a random guid rather than a specific client id. The client id doesn't seem to be necessary as the `MsalCacheHelper` only uses it to create a `PublicClientApplication` to iterate through the accounts in a serialized cache, in the method `GetAccountIdentifiersAsync`

https://github.com/AzureAD/microsoft-authentication-extensions-for-dotnet/blob/1a2de3a3ae272f288dbeda8fc242e7846892848f/src/Microsoft.Identity.Client.Extensions.Msal/MsalCacheHelper.cs#L91-L127

### PR DESCRIPTION
**Pull request by bgavrilMS**
Before this fix, developers had to configure client_id. 

This design does not take into account the following, which this PR fixes:

- it's not clear that `CacheChanged` event is only relevant for the given client_id
- since authority is not configured, the default authority is implied (i.e. `l.m.o/common`). This is the root cause of bug #89 
- developers cannot use 1 instance of the CacheExtension for multiple client_ids (since a client_id is needed to construct it)

Fixes #98 #123 #121 #102  (yay!)

### CODE REVIEW

File: sample/ManualTestApp/Program.cs [hunk 1]
```diff
@@ -276,8 +276,7 @@ private static async Task<MsalCacheHelper> CreateCacheHelperAsync()
             {
                 storageProperties = new StorageCreationPropertiesBuilder(
```
Comment (bgavrilMS) [inline sample/ManualTestApp/Program.cs:277]: Example usage is here - you can see that Client_ID is no longer a mandatory param. If people need CacheChanged event, they need to configure both "client_id" and "authority". If they don't, they'll get a good error message when subscribing to `CacheChanged`

File: sample/ManualTestApp/Program.cs [hunk 2]
```diff
@@ -287,6 +286,9 @@ private static async Task<MsalCacheHelper> CreateCacheHelperAsync()
                 .WithMacKeyChain(
                     Config.KeyChainServiceName,
                     Config.KeyChainAccountName)
+                .WithCacheChangedEvent( // do NOT use unless really necessary, high perf penalty!
```
Comment (bgavrilMS) [inline sample/ManualTestApp/Program.cs:289]: CacheChanged event is scoped to this (client_id, authority) pair. There isn't (currently) a way to get events about other pairs, except by creating different instances of the MsalCacheHelper object.

File: src/Shared/StorageCreationPropertiesBuilder.cs [hunk 3]
```diff
@@ -37,11 +38,25 @@ public class StorageCreationPropertiesBuilder
         /// <param name="cacheFileName">The name of the cache file to use when creating or opening storage.</param>
         /// <param name="cacheDirectory">The name of the directory containing the cache file.</param>
         /// <param name="clientId">The client id for the calling application</param>
+        [Obsolete("Use StorageCreationPropertiesBuilder(string, string) instead. " +
+            "If you need to consume the CacheChanged event then also use WithCacheChangedEvent(string, string)", false)]
         public StorageCreationPropertiesBuilder(string cacheFileName, string cacheDirectory, string clientId)
```
Comment (bgavrilMS) [inline src/Shared/StorageCreationPropertiesBuilder.cs:43]: This constructor is now obsolete.

**Comment by bgavrilMS** [comment]
Also adding @jmprieur for the public API changes. 
Note that I'll be merging this tomorrow.

File: sample/ManualTestApp/Program.cs (hunk 2 above)
Comment (jmprieur) [reply sample/ManualTestApp/Program.cs:289]: Does it mean that you won't have changes in the FOCI ?

**Review by jmprieur — COMMENTED** [review]
LGTM.
I don't fully understand how this would be a shared cache from the point of view of the change events?

**Comment by bgavrilMS** [comment]
@jmprieur - for FOCI apps, the client_id is not used in the internal logic of `GetAccounts`. Great question by the way. So no, there won't be any changes to VS or other FOCI apps.

Related to this being a shared cache. Think of this object as a singleton, as it represents a file on disk. I can then register a bunch of `ITokenCache` objects (from different `PublicClientApplication` objs) to synchronize with this file. If I'm interested in knowing which accounts were added / removed from the cache, I can register for an event: 

```csharp 
StoregePropertiesBuilder().WithCacheChangedEvent(client_id, authority);
```
and later subscribe to the event

```csharp
MsalCacheHelper helper = MsalCacheHelper.CreateAsync(storageProperties);
helper.CacheChangedEvent += <listner>
```

The `CacheChangeEvent` will only fire for the (client_id, authority) that you have configured. If you don't configure any, I'll throw.

You cannot currently listen to events for different pairs of (client_id, authority). VS doesn't care about it and nobody else expressed an interest in this event. As such, I think this is the simplest solution. 

A better solution would be to have an API like

```csharp
helper.RegisterCacheChangedCallback(client_id, authority, Action<accounts_added, accounts_removed>)
```

This is doable, but more work. Do you think it's worth doing now?

**Comment by jmprieur** [comment]
Thanks for the explanations, @bgavrilMS 
No, let's not do it until someone expresses their need.

File: src/Shared/StorageCreationPropertiesBuilder.cs [hunk 4]
```diff
@@ -79,6 +95,23 @@ public StorageCreationPropertiesBuilder WithMacKeyChain(string serviceName, stri
...
+        /// Enables the use of the MsalCacheHelper.CacheChanged event, which notifies about
+        /// accounts added and removed. These accounts are scoped to the client_id and authority
+        /// specified here.
+        /// </summary>
+        /// <param name="clientId">The client id for which you wish to receive notifications</param>
+        /// <param name="authority">The authority for which you wish to receive notifications</param>
+        /// <returns>The augmented builder</returns>        
+        public StorageCreationPropertiesBuilder WithCacheChangedEvent(
+            string clientId,
+            string authority = "https://login.microsoftonline.com/common")
```
Comment (henrik-me) [inline src/Shared/StorageCreationPropertiesBuilder.cs:108]: >"https://login.microsoftonline.com/common" [](start = 31, length = 42)

this is used in a number of places, should this be a const from MSAL, ensuring we only have one place we need to change default authority?

File: tests/Microsoft.Identity.Client.Extensions.Msal.UnitTests/MsalCacheHelperTests.cs [hunk 5]
```diff
@@ -393,7 +419,77 @@ public async Task EventFiresAsync()
             string cacheWithOneUser = File.ReadAllText(
                 ResourceHelper.GetTestResourceRelativePath("token_cache_one_acc_seed.json"));
 
+            var properties = _storageCreationPropertiesBuilder
+                .WithCacheChangedEvent(ClientId, "https://login.microsoftonline.com/common")
```
Comment (henrik-me) [inline tests/Microsoft.Identity.Client.Extensions.Msal.UnitTests/MsalCacheHelperTests.cs:423]: >"https://login.microsoftonline.com/common" [](start = 49, length = 42)

any reason this isn't a const like ClientID ?

**Review by henrik-me — APPROVED** [review]
:shipit:

**Pull request merged**